File: src/lib/effort.ts
// Indicateur d'effort d'une séance, façon Strava (« Facile » → « Maximal »).
// S'appuie sur la zone cardio moyenne si une FC est disponible, sinon retombe
// sur la durée. Estimation indicative, pas une mesure physiologique.
import { heartRateZone } from '@/lib/calories';
import type { Session } from '@/lib/types';

export type EffortLevel = 1 | 2 | 3 | 4 | 5;

export type Effort = {
  level: EffortLevel;
  label: string;
  /** Clé de couleur du thème (data viz : du vert au rose cardio). */
  colorKey: 'success' | 'warning' | 'heart';
};

const LABELS: Record<EffortLevel, string> = {
  1: 'Facile',
  2: 'Modéré',
  3: 'Soutenu',
  4: 'Difficile',
  5: 'Maximal',
};

// Vert pour le facile, ambre pour le soutenu, rose cardio pour le maximal.
const COLORS: Record<EffortLevel, Effort['colorKey']> = {
  1: 'success',
  2: 'success',
  3: 'warning',
  4: 'warning',
  5: 'heart',
};

export function sessionEffort(session: Session, maxHr: number): Effort {
  let level: EffortLevel;
  if (session.avgHr != null && maxHr > 0) {
    level = heartRateZone(session.avgHr, maxHr) as EffortLevel;
  } else {
    // Hors FC, l'effort suit la durée : on prend le temps en mouvement quand il
    // est connu (vélo) pour qu'un long arrêt ne le surévalue pas.
    const minutes = (session.movingTimeSec ?? session.durationSec) / 60;
    level = minutes < 30 ? 1 : minutes < 60 ? 2 : minutes < 90 ? 3 : minutes < 150 ? 4 : 5;
  }
  return { level, label: LABELS[level], colorKey: COLORS[level] };
}

File: app/src/main/java/ovh/battistella/elan/domain/Effort.kt
// Indicateur d'effort d'une séance, façon Strava (« Facile » → « Maximal »).
// S'appuie sur la zone cardio moyenne si une FC est disponible, sinon retombe
// sur la durée. Estimation indicative, pas une mesure physiologique.
package ovh.battistella.elan.domain

data class Effort(
    /** Niveau de 1 à 5. */
    val level: Int,
    val label: String,
    /** Clé de couleur du thème : `success`, `warning` ou `heart`. */
    val colorKey: String,
)

private val LABELS = mapOf(1 to "Facile", 2 to "Modéré", 3 to "Soutenu", 4 to "Difficile", 5 to "Maximal")

// Vert pour le facile, ambre pour le soutenu, rose cardio pour le maximal.
private val COLORS = mapOf(1 to "success", 2 to "success", 3 to "warning", 4 to "warning", 5 to "heart")

fun sessionEffort(session: Session, maxHr: Double): Effort {
    val avgHr = session.avgHr
    val level = if (avgHr != null && maxHr > 0) {
        heartRateZone(avgHr, maxHr)
    } else {
        // Hors FC, l'effort suit la durée : temps en mouvement quand il est connu,
        // pour qu'un long arrêt ne le surévalue pas.
        val minutes = (session.movingTimeSec ?: session.durationSec) / 60.0
        when {
            minutes < 30 -> 1
            minutes < 60 -> 2
            minutes < 90 -> 3
            minutes < 150 -> 4
            else -> 5
        }
    }
    return Effort(level = level, label = LABELS.getValue(level), colorKey = COLORS.getValue(level))
}
